kernel/comps/virtio/src/device/input/device.rs

```diff
@@ -179,13 +185,67 @@ impl InputDevice {
         let size = field_ptr!(&self.config, VirtioInputConfig, size)
             .read_once()
             .unwrap();
-        let data: [u8; 128] = field_ptr!(&self.config, VirtioInputConfig, data)
-            // FIXME: It is impossible to call `read_once` on `[u8; 128]`. What's the proper way to
-            // read this field out?
-            .read()
-            .unwrap();
-        out[..size as usize].copy_from_slice(&data[..size as usize]);
-        size
+        let data_offset = offset_of!(VirtioInputConfig, data);
+        let mut data_ptr = self.config.borrow_vm();
+        data_ptr.byte_add(data_offset as usize);
+
+        match select {
+            InputConfigSelect::IdName
+            | InputConfigSelect::IdSerial
+            | InputConfigSelect::PropBits
+            | InputConfigSelect::EvBits => {
```

Comment:
Could you just split the method into multiple ones? For example:
```rust
    pub fn query_config_id_name(&self) -> String {
        let size = self.select_config(InputConfigSelect::IdName, 0);

        // Do the actual reading...
    }

    pub fn query_config_prop_bits(&self) -> InputProp {
        let size = self.select_config(InputConfigSelect::PropBits, 0);

        // Check `size == 1`, read one byte, and construct `InputProp`...
    }

    fn select_config(&self, select: InputConfigSelect, subsel: u8) -> usize {
        field_ptr!(&self.config, VirtioInputConfig, select)
            .write_once(&(select as u8))
            .unwrap();
        field_ptr!(&self.config, VirtioInputConfig, subsel)
            .write_once(&subsel)
            .unwrap();
        field_ptr!(&self.config, VirtioInputConfig, size)
            .read_once()
            .unwrap()
    }
```

This will allow us to construct `AbsInfo` directly, instead of first constructing `[u8; 128]` and then converting it to `AbsInfo`. (It seems that no one is actually querying `AbsInfo` right now, so please don't add the method to avoid adding dead code. We can do this in the future).

Reply:
Done. BTW, the `size` will be 0 if the `prop` is empty.

Reply:
> the `size` will be 0 if the `prop` is empty.

Interesting. This seems to mean that our previous code is actually buggy:
https://github.com/asterinas/asterinas/blob/4823b82e4164e6aa4f58a982be8b27c07191b8d9/kernel/comps/virtio/src/device/input/device.rs#L120-L123